File: 2020/9.php
<?php

require('libs/core.php');

function check_1($numbers, $max) {
    $array = array_slice($numbers, 0, $max);
    $i = $max;
    while ($i < count($numbers)) {
        $found = false;
        foreach ($array as $l1 => $n1) {
            foreach ($array as $l2 => $n2) {
                if (($n1 + $n2 == $numbers[$i]) && ($l1 != $l2)) {
                    $array[] = $numbers[$i];
                    $array = array_slice($array, 1);
                    $found = true;
                    break 2;
                }
            }
        }
        if (!$found) {
            print("Could not find {$numbers[$i]}\n");
            return $numbers[$i];
        }
        $i++;
    }
}

function check_2($numbers, $invalid) {
    for ($i=1; $i < count($numbers); $i++) {
        $array = [];
        for ($j = $i; $j > 0; $j--) {
            $array[] = $numbers[$j];
            $sum = array_sum($array);
            if ($sum == $invalid) {
                sort($array);
                $max = end($array);
                $min = reset($array);
                print "Sum is " . ($min + $max) . "\n";
                return;
            }
            if ($sum > $invalid) {
                break;
            }
        }
    }
}

$numbers = load_data("9.txt");

$invalid = check_1($numbers, 25);

check_2($numbers, $invalid);

File: 2020/python/9.py
import core
import re

def check_1(numbers, max):
	array = numbers[0:max]
	i = max
	count = len(numbers)
	while i < count:
		found = False
		for l1, n1 in enumerate(array):
			for l2, n2 in enumerate(array):
				if (found == False) and (int(n1) + int(n2) == int(numbers[i])) and (l1 != l2):
					array.append(numbers[i])
					array = array[1:]
					found = True
					break
		if found == False:
			print("Could not find %d" % int(numbers[i]))
			return int(numbers[i])
		i += 1

def check_2(numbers, invalid):
	for i, value in enumerate(numbers):
		array = []
		for j in range(i, 0, -1):
			array.append(int(numbers[j]))
			count = sum(array)
			if count == invalid:
				array.sort()
				max = array[-1]
				min = array[0]
				print("Sum is %d" % (min + max))
				return
			if count > invalid:
				break

numbers = core.load_data("9.txt")

invalid = check_1(numbers, 25)

check_2(numbers, invalid)
